v-next/hardhat-viem/src/contracts.ts

```diff
@@ -0,0 +1,233 @@
+import type {
+  ContractReturnType,
+  DeployContractConfig,
+  GetContractAtConfig,
+  GetTransactionReturnType,
+  Libraries,
+  PublicClient,
+  SendDeploymentTransactionConfig,
+  WalletClient,
+} from "./types.js";
+import type { ArtifactsManager } from "@ignored/hardhat-vnext/types/artifacts";
+import type { EthereumProvider } from "@ignored/hardhat-vnext/types/providers";
+import type { PrefixedHexString } from "@ignored/hardhat-vnext-utils/hex";
+import type * as viemT from "viem";
+
+import { HardhatError } from "@ignored/hardhat-vnext-errors";
+import { toBigInt } from "@ignored/hardhat-vnext-utils/bigint";
+import { resolveLinkedBytecode } from "@ignored/hardhat-vnext-utils/bytecode";
+import { ensureError } from "@ignored/hardhat-vnext-utils/error";
+import { getContractAddress, getContract } from "viem";
+
+import { getDefaultWalletClient, getPublicClient } from "./clients.js";
+
+export async function deployContract<ContractName extends string>(
+  provider: EthereumProvider,
+  artifactManager: ArtifactsManager,
+  contractName: ContractName,
+  constructorArgs: unknown[] = [],
+  deployContractConfig: DeployContractConfig = {},
+): Promise<ContractReturnType<ContractName>> {
+  const {
+    client,
+    confirmations = 1,
+    libraries = {},
+    ...deployContractParameters
+  } = deployContractConfig;
+
+  if (confirmations < 0) {
```

Comment:
This used to be placed lower in the code, but since it’s just a parameter check, it's better to perform the validation earlier. This way, we can fail faster in case of an error.